plugin_test.go

```diff
@@ -0,0 +1,196 @@
+	}
+	if !bytes.Equal(got, want) {
+		t.Errorf("plugin() wrote unexpected secret value. got = %v, want = %v", got, want)
+	}
+}
+
+func TestPluginSMError(t *testing.T) {
+	dir, cleanup := driveMountHelper(t)
+	defer cleanup()
+
+	client, shutdown := mock(t, &mockSecretServer{
+		accessFn: func(ctx context.Context, _ *secretmanagerpb.AccessSecretVersionRequest) (*secretmanagerpb.AccessSecretVersionResponse, error) {
+			return nil, status.Error(codes.FailedPrecondition, "Secret is Disabled")
+		},
+	})
+	defer shutdown()
+
+	got := plugin(context.Background(), client, sampleAttrs, "{}", dir, "777")
+	if !strings.Contains(got.Error(), "FailedPrecondition") {
+		t.Errorf("plugin() got err = %v, want err = nil", got)
+
+	}
+}
+
+func TestPluginConfigErrors(t *testing.T) {
+	dir, cleanup := driveMountHelper(t)
+	defer cleanup()
+
+	client, shutdown := mock(t, &mockSecretServer{
+		accessFn: func(ctx context.Context, _ *secretmanagerpb.AccessSecretVersionRequest) (*secretmanagerpb.AccessSecretVersionResponse, error) {
+			return &secretmanagerpb.AccessSecretVersionResponse{
+				Payload: &secretmanagerpb.SecretPayload{
+					Data: []byte("data"),
+				},
+			}, nil
+		},
+	})
+	defer shutdown()
+
+	tests := []struct {
+		name    string
+		attrs   string
+		secrets string
+		dir     string
+		perm    string
+	}{
+		{
+			name:    "unparsable attributes",
+			attrs:   "",
+			secrets: "{}",
+			dir:     dir,
+			perm:    "777",
+		},
+		{
+			name:    "unparsable secrets",
+			attrs:   sampleAttrs,
+			secrets: "",
+			dir:     dir,
+			perm:    "777",
+		},
+		{
+			name:    "unparsable perm",
+			attrs:   sampleAttrs,
+			secrets: "{}",
+			dir:     dir,
+			perm:    "foobar",
+		},
+	}
+
+	for _, tc := range tests {
+		t.Run(tc.name, func(t *testing.T) {
+			got := plugin(context.Background(), client, tc.attrs, tc.secrets, tc.dir, tc.perm)
+			if got == nil {
+				t.Errorf("plugin() succeeded for malformed input, want error")
+			}
+		})
+	}
+}
+
+// driveMountHelper creates a temporary directory for use by tests as a
+// replacement for the tmpfs directory that the CSI Driver would create for the
+// plugin. This returns the path and a cleanup function to run at the end of
+// the test.
+func driveMountHelper(t *testing.T) (string, func()) {
+	t.Helper()
+	dir, err := ioutil.TempDir("", "csi-test")
+	if err != nil {
+		t.Fatal(err)
+	}
+	return dir, func() {
+		log.Printf("Cleaning up: %s", dir)
+		os.RemoveAll(dir)
+	}
+}
+
+// mock builds a secretmanager.Client talking to a real in-memory secretmanager
+// GRPC server of the *mockSecretServer. It also returns a function that must
+// be called to shut down the grpc server gracefully.
+func mock(t *testing.T, m *mockSecretServer) (*secretmanager.Client, func()) {
+	t.Helper()
+	l := bufconn.Listen(1024 * 1024)
+	s := grpc.NewServer()
+	secretmanagerpb.RegisterSecretManagerServiceServer(s, m)
+
+	go func() {
+		if err := s.Serve(l); err != nil {
```

Comment:
Prolly want to ignore `grpc.ErrServerStopped`

Reply:
```
// Serve will return a non-nil error unless Stop or GracefulStop is called.
```

Since we _should_ always call `Stop` this should never return an error in the good case so I dont think ignoring that error is necessary.

Reply:
Hmmmmm. I think that comment is wrong from my testing.